src/stream/src/executor/arrangement_backfill.rs

```diff
@@ -0,0 +1,679 @@
+        // They contain the backfill position,
+        // as well as the progress.
+        // However, they do not contain the vnode key at index 0.
+        // That is filled in when we flush the state table.
+        let mut current_state: Vec<Datum> = vec![None; state_len];
+        let mut old_state: Option<Vec<Datum>> = None;
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(first_barrier);
+
+        // If no need backfill, but state was still "unfinished" we need to finish it.
+        // So we just update the state + progress to meta at the next barrier to finish progress,
+        // and forward other messages.
+        //
+        // Reason for persisting on second barrier rather than first:
+        // We can't update meta with progress as finished until state_table
+        // has been updated.
+        // We also can't update state_table in first epoch, since state_table
+        // expects to have been initialized in previous epoch.
+
+        // The epoch used to snapshot read upstream mv.
+        let mut snapshot_read_epoch;
+
+        // Keep track of rows from the snapshot.
+        let mut total_snapshot_processed_rows: u64 = 0;
+
+        // Backfill Algorithm:
+        //
+        //   backfill_stream
+        //  /               \
+        // upstream       snapshot
+        //
+        // We construct a backfill stream with upstream as its left input and mv snapshot read
+        // stream as its right input. When a chunk comes from upstream, we will buffer it.
+        //
+        // When a barrier comes from upstream:
+        //  - Update the `snapshot_read_epoch`.
+        //  - For each row of the upstream chunk buffer, forward it to downstream if its pk <=
+        //    `current_pos`, otherwise ignore it.
+        //  - reconstruct the whole backfill stream with upstream and new mv snapshot read stream
+        //    with the `snapshot_read_epoch`.
+        //
+        // When a chunk comes from snapshot, we forward it to the downstream and raise
+        // `current_pos`.
+        //
+        // When we reach the end of the snapshot read stream, it means backfill has been
+        // finished.
+        //
+        // Once the backfill loop ends, we forward the upstream directly to the downstream.
+        if to_backfill {
+            let mut upstream_chunk_buffer: Vec<StreamChunk> = vec![];
+            let mut pending_barrier: Option<Barrier> = None;
+            'backfill_loop: loop {
+                let mut cur_barrier_snapshot_processed_rows: u64 = 0;
+                let mut cur_barrier_upstream_processed_rows: u64 = 0;
+
+                // Process barrier
+                // Each time we break out of the backfill_stream loop,
+                // it is typically due to barrier (except the first time).
+                // So we should immediately process barrier if there's one pending.
+                if let Some(barrier) = pending_barrier.clone() {
```

Comment:
Barrier processing logic has been shifted inside this block.

`if let Some(barrier) = ...` then process barrier.